apps/aecore/src/aec_monitor_metrics_probe.erl

```diff
@@ -0,0 +1,165 @@
+-module(aec_monitor_metrics_probe).
+-behavior(exometer_probe).
+
+-export([behaviour/0,
+         probe_init/3,
+         probe_terminate/1,
+         probe_setopts/3,
+         probe_update/2,
+         probe_get_value/2,
+         probe_get_datapoints/1,
+         probe_reset/1,
+         probe_sample/1,
+         probe_handle_msg/2,
+         probe_code_change/3]).
+
+-export([ad_hoc_spec/0]).
+
+
+-record(s,
+        {updates  = 0            :: non_neg_integer(),
+         stats    = init_stats() :: #{[atom()] := integer()}}).
+
+
+ad_hoc_spec() ->
+    [{module, ?MODULE},
+     {type, probe},
+     {sample_interval, infinity}].
+
+
+-spec behaviour() -> exometer:behaviour().
+
+behaviour() -> probe.
+
+
+%%% Lifecycle
+
+probe_init(Name, Type, Options) ->
+    ok = lager:info("~p:probe_init(~p, ~p, ~p)", [?MODULE, Name, Type, Options]),
+    _ = erlang:send_after(2000, self(), update),
+    Stats = init_stats(),
+    ok = ensure_metrics(Stats),
+    ok = set_triggers(),
+    {ok, #s{stats = Stats}}.
+
+set_triggers() ->
+    Triggers =
+        [{long_gc, 500},
+         {long_schedule, 500},
+         {large_heap, 1024 * 1024},
+         busy_port,
+         busy_dist_port],
+    _ = erlang:system_monitor(self(), Triggers),
+    ok.
+
+
+probe_terminate(Reason) ->
+    ok = lager:debug("~p terminating: ~p", [?MODULE, Reason]),
+    ok.
+
+
+%%% External interface
+
+probe_get_value(default, State = #s{updates = Count}) ->
+    {ok, [{updates, Count}], State};
+probe_get_value(Keys, State) ->
+    Values = maps:to_list(maps:with(Keys, all_datapoints(State))),
+    {ok, Values, State}.
+
+
+probe_get_datapoints(State) ->
+    Datapoints = maps:to_list(all_datapoints(State)),
+    {ok, Datapoints}.
+
+
+probe_update(_, _) ->
+    {error, not_supported}.
+
+
+probe_reset(State) ->
+    NewState = State#s{updates = 0, stats = init_stats()},
+    {ok, NewState}.
+
+
+probe_sample(_) ->
+    {error, not_supported}.
+
+
+probe_setopts(_Entry, Opts, State = #s{stats = Stats}) ->
+    Keys = proplists:get_value(datapoints, Opts, maps:keys(Stats)),
+    NewStats = maps:with(Keys, Stats),
+    NewState = State#s{stats = NewStats},
+    {ok, NewState}.
+
+
+probe_handle_msg(update, State) ->
+    _ = erlang:send_after(2000, self(), update),
+    NewState = do_update(State),
+    {ok, NewState};
+probe_handle_msg({monitor, PID, Tag, Info}, State) ->
+    NewState = handle_monitor(PID, Tag, Info, State),
+    {ok, NewState};
+probe_handle_msg(Unexpected, State) ->
+    ok = lager:debug("Unknown msg: ~p", [Unexpected]),
+    {ok, State}.
+
+
+probe_code_change(_, State, _) ->
+    {ok, State}.
+
+
+do_update(State = #s{stats = Stats}) ->
+    NewStats = maps:merge(Stats, stats()),
+    NewState = State#s{stats = NewStats},
+    Datapoints = maps:to_list(all_datapoints(NewState)),
+    NotifyMetrics = fun({N, V}) -> aec_metrics:try_update(N, V) end,
+    ok = lists:foreach(NotifyMetrics, Datapoints),
```

Comment:
Haha, indeed. I think I like the second version better. Just didn't cross my mind -- which is funny because that's actually the style I prefer! I don't particularly like making lambdas that don't close over anything.